File: packages/server/src/speech/service.ts
import { z } from "zod"
import type { Readable } from "node:stream"
import type { Logger } from "../logger"
import type { SettingsService } from "../settings/service"
import type { SpeechCapabilitiesResponse, SpeechSynthesisResponse, SpeechTranscriptionResponse } from "../api-types"
import { OpenAICompatibleSpeechProvider } from "./providers/openai-compatible"

const ServerSpeechSettingsSchema = z.object({
  speech: z
    .object({
      provider: z.string().optional(),
      apiKey: z.string().optional(),
      baseUrl: z.string().optional(),
      sttModel: z.string().optional(),
      ttsModel: z.string().optional(),
      ttsVoice: z.string().optional(),
      ttsFormat: z.enum(["mp3", "wav", "opus", "aac"]).optional(),
    })
    .optional(),
})

export interface TranscribeAudioInput {
  audioBase64: string
  mimeType: string
  filename?: string
  language?: string
  prompt?: string
}

export interface SynthesizeSpeechInput {
  text: string
  format?: "mp3" | "wav" | "opus" | "aac"
}

export interface SpeechSynthesisStreamResponse {
  stream: Readable
  mimeType: string
}

export interface SpeechProvider {
  getCapabilities(): SpeechCapabilitiesResponse
  transcribe(input: TranscribeAudioInput): Promise<SpeechTranscriptionResponse>
  synthesize(input: SynthesizeSpeechInput): Promise<SpeechSynthesisResponse>
  synthesizeStream(input: SynthesizeSpeechInput): Promise<SpeechSynthesisStreamResponse>
}

export interface NormalizedSpeechSettings {
  provider: string
  apiKey?: string
  baseUrl?: string
  sttModel: string
  ttsModel: string
  ttsVoice: string
  ttsFormat: "mp3" | "wav" | "opus" | "aac"
}

const DEFAULT_PROVIDER = "openai-compatible"
const DEFAULT_STT_MODEL = "gpt-4o-mini-transcribe"
const DEFAULT_TTS_MODEL = "gpt-4o-mini-tts"
const DEFAULT_TTS_VOICE = "alloy"
const DEFAULT_TTS_FORMAT = "mp3"
export class SpeechService {
  constructor(
    private readonly settings: SettingsService,
    private readonly logger: Logger,
  ) {}

  getCapabilities(): SpeechCapabilitiesResponse {
    return this.createProvider().getCapabilities()
  }

  async transcribe(input: TranscribeAudioInput): Promise<SpeechTranscriptionResponse> {
    return this.createProvider().transcribe(input)
  }

  async synthesize(input: SynthesizeSpeechInput): Promise<SpeechSynthesisResponse> {
    return this.createProvider().synthesize(input)
  }

  async synthesizeStream(input: SynthesizeSpeechInput): Promise<SpeechSynthesisStreamResponse> {
    return this.createProvider().synthesizeStream(input)
  }

  private createProvider(): SpeechProvider {
    const settings = this.resolveSettings()
    return new OpenAICompatibleSpeechProvider({
      settings,
      logger: this.logger.child({ provider: settings.provider }),
    })
  }

  private resolveSettings(): NormalizedSpeechSettings {
    const parsed = ServerSpeechSettingsSchema.parse(this.settings.getOwner("config", "server") ?? {})
    const speech = parsed.speech ?? {}

    return {
      provider: speech.provider?.trim() || DEFAULT_PROVIDER,
      apiKey: speech.apiKey?.trim() || process.env.OPENAI_API_KEY,
      baseUrl: speech.baseUrl?.trim() || process.env.OPENAI_BASE_URL || undefined,
      sttModel: speech.sttModel?.trim() || DEFAULT_STT_MODEL,
      ttsModel: speech.ttsModel?.trim() || DEFAULT_TTS_MODEL,
      ttsVoice: speech.ttsVoice?.trim() || DEFAULT_TTS_VOICE,
      ttsFormat: speech.ttsFormat ?? DEFAULT_TTS_FORMAT,
    }
  }
}

File: packages/server-rust/src/speech/service.rs
use async_trait::async_trait;
use crate::api_types::{SpeechCapabilitiesResponse, SpeechSynthesisResponse, SpeechTranscriptionResponse};
use crate::logger::Logger;
use crate::settings::service::{DocKind, SettingsService};
use crate::speech::providers::openai_compatible::OpenAICompatibleSpeechProvider;

const DEFAULT_PROVIDER: &str = "openai-compatible";
const DEFAULT_STT_MODEL: &str = "gpt-4o-mini-transcribe";
const DEFAULT_TTS_MODEL: &str = "gpt-4o-mini-tts";
const DEFAULT_TTS_VOICE: &str = "alloy";
const DEFAULT_TTS_FORMAT: &str = "mp3";

#[derive(Debug, Clone)]
pub struct TranscribeAudioInput {
    pub audio_base64: String,
    pub mime_type: String,
    pub filename: Option<String>,
    pub language: Option<String>,
    pub prompt: Option<String>,
}

#[derive(Debug, Clone)]
pub struct SynthesizeSpeechInput {
    pub text: String,
    pub format: Option<String>,
}

#[derive(Debug, Clone)]
pub struct NormalizedSpeechSettings {
    pub provider: String,
    pub api_key: Option<String>,
    pub base_url: Option<String>,
    pub stt_model: String,
    pub tts_model: String,
    pub tts_voice: String,
    pub tts_format: String,
}

#[async_trait]
pub trait SpeechProvider {
    fn get_capabilities(&self) -> SpeechCapabilitiesResponse;
    async fn transcribe(&self, input: TranscribeAudioInput) -> Result<SpeechTranscriptionResponse, String>;
    async fn synthesize(&self, input: SynthesizeSpeechInput) -> Result<SpeechSynthesisResponse, String>;
}

pub struct SpeechService {
    settings: SettingsService,
    logger: Logger,
}

impl SpeechService {
    pub fn new(settings: SettingsService, logger: Logger) -> Self {
        Self { settings, logger }
    }

    pub fn get_capabilities(&mut self) -> SpeechCapabilitiesResponse {
        self.create_provider().get_capabilities()
    }

    pub async fn transcribe(&mut self, input: TranscribeAudioInput) -> Result<SpeechTranscriptionResponse, String> {
        self.create_provider().transcribe(input).await
    }

    pub async fn synthesize(&mut self, input: SynthesizeSpeechInput) -> Result<SpeechSynthesisResponse, String> {
        self.create_provider().synthesize(input).await
    }

    fn create_provider(&mut self) -> OpenAICompatibleSpeechProvider {
        let settings = self.resolve_settings();
        let provider_name = settings.provider.clone();
        OpenAICompatibleSpeechProvider::new(
            settings,
            self.logger.child(&provider_name),
        )
    }

    fn resolve_settings(&mut self) -> NormalizedSpeechSettings {
        let server = self.settings.get_owner(&DocKind::Config, "server");
        let speech = server.get("speech");

        let provider = speech
            .and_then(|s| s.get("provider"))
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| DEFAULT_PROVIDER.to_string());

        let api_key = speech
            .and_then(|s| s.get("apiKey"))
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .or_else(|| std::env::var("OPENAI_API_KEY").ok());

        let base_url = speech
            .and_then(|s| s.get("baseUrl"))
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .or_else(|| std::env::var("OPENAI_BASE_URL").ok());

        let stt_model = speech
            .and_then(|s| s.get("sttModel"))
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| DEFAULT_STT_MODEL.to_string());

        let tts_model = speech
            .and_then(|s| s.get("ttsModel"))
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| DEFAULT_TTS_MODEL.to_string());

        let tts_voice = speech
            .and_then(|s| s.get("ttsVoice"))
            .and_then(|v| v.as_str())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| DEFAULT_TTS_VOICE.to_string());

        let tts_format = speech
            .and_then(|s| s.get("ttsFormat"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| DEFAULT_TTS_FORMAT.to_string());

        NormalizedSpeechSettings {
            provider,
            api_key,
            base_url,
            stt_model,
            tts_model,
            tts_voice,
            tts_format,
        }
    }
}
